Multi-country targeting campaign - Integrate UI with `ads/campaigns` related endpoints

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implemented the "**Integrate `ads/campaigns` related endpoints with mocking API**" part of #1248, and it's based on #1271.

- Adapt each received campaign entity in *actions.js*.
- Change the `country` initial value to `countryCodes` for all campaign \<Form\> and update related uses.
- Change campaign creation to multiple countries.
- Show multiple countries on `<AllProgramsTableCard>`.

💡  Please note that the UI integration of budget section will be done in another subsequence PR.

### Screenshots:

#### 📹  Create campaign
https://user-images.githubusercontent.com/17420811/155689353-73ced7d1-4cf7-4412-ab39-5629b68fbe00.mp4

#### 📷  Edit multi-country targeting campaign
![image](https://user-images.githubusercontent.com/17420811/155688647-f6f16ff5-2480-4890-8925-8818ff70d31a.png)

#### 📷  Programs table
![image](https://user-images.githubusercontent.com/17420811/155689591-017ffb25-2e05-4a68-9078-a5b54fb59e2c.png)

#### Event tracking of `gla_launch_paid_campaign_button_click`
![image](https://user-images.githubusercontent.com/17420811/155689338-744a5ac6-2a48-4d0d-9987-2ad0f7f14a1d.png)

### Detailed test instructions:

💡  Please note that this PR is currently using mocked API. And the country list used in `<AudienceCountrySelect>` is the selected targeting audience when setting free listings (or onboarding). So that the mocking countries in the `targeted_locations` of campaign would not appear on the selector UI if they are not in your targeting audience.
https://github.com/woocommerce/google-listings-and-ads/blob/f3f4e1e5c82789bed7b4c8692a76c77031c49303/js/src/components/audience-country-select.js#L8-L15
1. Open browser DevTool and run `localStorage.setItem( 'debug', 'wc-admin:*' )` in the Console tab for checking the event tracking.
2. Go to create campaign and check if the country selector UI works correctly.
3. Finish the creation. Check if the `audiences` property is shown in the format of concatenating country codes with `','`. For example `'US,JP,AU'`.
4. With at least two campaigns, go to dashboard to check if the multiple countries are shown in "United States (US) + 2 more" format in the **Country** column of Programes table.
5. Edit a multiple countries campaign and check if the country selector UI is displayed correctly.
6. Edit a single countries campaign and check if the country selector UI is displayed correctly.

### Changelog entry

eason9487 committed:
Adapte campaign entity in actions.js and add some JSDocs

eason9487 committed:
Change the `country` initial value to `countryCodes` for all <Form> and update related uses.

eason9487 committed:
Change to create new campaign with multiple countries.

eason9487 committed:
Show countries in AllProgramsTableCard

eason9487 committed:
Add `displayCountries` when adapting received campaign entity for reducing repeated codes.

puntope commented on js/src/dashboard/all-programs-table-card/index.js:
```diff
@@ -44,6 +44,21 @@ const headers = [
+			{ countryNameMap[ first ] }
+			{ countryCodes.length >= 2 &&
+				sprintf(
+					// translators: %s: number of countries, with minimum value of 1.
+					__( ' + %s more', 'google-listings-and-ads' ),
```
💅 Checking [the repo](https://www.npmjs.com/package/sprintf-js#:~:text=that%20ASCII%20value-,d,-or%20i%20%E2%80%94%20yields)

I'd say this should be %d

```suggestion
					// translators: %d: number of countries, with minimum value of 1.
					__( ' + %d more', 'google-listings-and-ads' ),
```

eason9487 replied:
Thanks for catching this! 👍  Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1273/commits/188907e30e8eb401f7a6b044ae639d3e1c67c129.

puntope commented on js/src/dashboard/all-programs-table-card/index.js:
```diff
@@ -44,6 +44,21 @@ const headers = [
 	{ key: 'actions', label: '', required: true },
 ];
 
+function CountryColumn( { countryCodes, countryNameMap } ) {
```
💅  Having Components in independent files reduces coupling IMO.

eason9487 replied:
Their relationship should be more likely to be high cohesion. Making it a separate file probably does not make much difference in terms of coupling.

puntope commented on js/src/data/actions.js:
```diff
@@ -740,13 +760,34 @@ export function* saveTargetAudience( targetAudience ) {
 	}
 }
 
+/**
+ * Adaptes the campaign entity received from API.
+ *
+ * @param {Object} campaign The campaign entity to be adapted.
+ * @return {Campaign} Campaign data.
+ */
+function adapteAdsCampaign( campaign ) {
+	const allowMultiple = campaign.targeted_locations.length > 0;
+	const displayCountries = allowMultiple
+		? campaign.targeted_locations
+		: [ campaign.country ];
+	return {
+		...campaign,
+		allowMultiple,
+		displayCountries,
+	};
+}
```
💅  Until now in the `actions.js` file we only have the action functions. Now we also have this auxiliary function for mapping the Ads campaign. Now it's only one. But wondering if in the function we will start adding more and more aux functions together with the actions. 

I'd suggest to have it in an external file

eason9487 replied:
Good idea 💯 . Moved in https://github.com/woocommerce/google-listings-and-ads/pull/1273/commits/593dc2e0ab05588aef0b4821ff4c649d8715ef36 with fixing a typo. 😅

puntope commented on src/Tracking/README.md:
```diff
@@ -149,7 +149,8 @@ All event names are prefixed by `wcadmin_gla_`.
 -   `launch_paid_campaign_button_click` - Triggered when the "Launch paid campaign" button is clicked to add a new paid campaign
 
-    -   `audience`: country code of the paid campaign audience country
+    -   `audience`: country code of the paid campaign audience country. This is inapplicable when a campaign is created with the multi-country targeting feature.
```
❓  We have actually this possibility? I mean to have only `audience`

eason9487 replied:
After introducing the multi-country targeting feature, the `audience` property will no longer appear in the tracking data. I kept it as a record in this README so that when looking at older tracking data, this property can still have a reference. Otherwise, there's no other document about it.

puntope replied:
Oh, gotcha. Is this how we do this normally? Wondering if would be better to just have it 

`audiences : country codes of the paid campaign audience countries, e.g. 'US,JP,AU'. Before multi-country support, it was implemented as 'audience'`

or even better IMO

`audience : country codes of the paid campaign audience countries, e.g. US,JP,AU'`

Maintaining the same name. Since "Audience" in the singular form also evokes a group, it makes sense also in naming.

puntope replied:
WDYT?

eason9487 replied:
> After introducing the multi-country targeting feature, the `audience` property will no longer appear in the tracking data.

Sorry I'm not quite right here. Even after the feature is released, there will still be merchants who are staying with old versions or haven't updated it yet, so the `audience` property will still be in the tracking data for quite some time.

> or even better IMO
> 
> `audience : country codes of the paid campaign audience countries, e.g. US,JP,AU'`
> 
> Maintaining the same name. Since "Audience" in the singular form also evokes a group, it makes sense also in naming.

Using the same property name will result in the following two types of data being identical, and it will be hard to identify which campaign is which:
- The old single-country targeting selected `US`
- The new multi-country targeting also selects only one country `US`

I updated their descriptions in https://github.com/woocommerce/google-listings-and-ads/pull/1273/commits/898c016f7e4dfd871c6192a2b5f641e4f5aac811, hope it could be more clear.

puntope reviewed (approved):
✅  LGTM

- Tested the UI, adding multiple countries, playing with the budget, removing countries etc.
- I could create the campaign 
- When I edited the campaign it was disabled showing a message. 
- They show the correct tracking in the console
- I see +2 more (in my case) in the Paid Campaigns table

💅  Some very minor comments regarding mainly naming

eason9487 committed:
Fix the format placeholder in all-programs-table-card/index.js

Address https://github.com/woocommerce/google-listings-and-ads/pull/1273#discussion_r816672854

eason9487 committed:
Move adaptAdsCampaign function out from actions.js and fix typo

Address https://github.com/woocommerce/google-listings-and-ads/pull/1273#discussion_r816701469

eason9487 committed:
Update descriptions of `audience` and `audiences` to make them more clear.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1273#discussion_r816709210
